File: GuildProfiles/src/main/java/xyz/wingio/plugins/guildprofiles/pages/Adapter.java
package xyz.wingio.plugins.guildprofiles.pages;

import android.content.Context;
import android.view.*;
import android.widget.RelativeLayout;
import android.graphics.*;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RectShape;
import android.util.Base64;
import android.app.*;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.annotation.DimenRes;
import androidx.fragment.app.FragmentActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.aliucord.Utils;
import com.aliucord.utils.*;
import com.aliucord.Http;
import com.aliucord.Logger;
import xyz.wingio.plugins.guildprofiles.util.*;
import com.aliucord.fragments.SettingsPage;
import com.discord.models.member.GuildMember;
import com.discord.models.user.User;
import com.discord.utilities.color.ColorCompat;
import com.discord.utilities.extensions.SimpleDraweeViewExtensionsKt;
import com.discord.utilities.icon.IconUtils;
import com.discord.utilities.images.MGImages;
import com.discord.stores.*;
import com.discord.widgets.user.usersheet.WidgetUserSheet;
import com.lytefast.flexinput.R;
import com.facebook.drawee.view.SimpleDraweeView;

import kotlin.jvm.functions.Function1;
import java.util.List;

import java.io.*;

public class Adapter extends RecyclerView.Adapter<ViewHolder> {
    private static final int layoutId = Utils.getResId("widget_user_profile_adapter_item_server", "layout");

    private final List<GuildMember> friends;
    private final SettingsPage page;
    public Logger logger = new Logger("GP");

    public Adapter(SettingsPage page, List<GuildMember> friends) {
        this.friends = friends;
        this.page = page;
    }

    @Override
    public int getItemCount() {
        return friends.size();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        var layout = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
        return new ViewHolder(this, (ConstraintLayout) layout);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        var friend = friends.get(position);
        var friendUser = StoreStream.getUsers().getUsers().get(friend.getUserId());
        var color = Integer.valueOf(ColorCompat.getThemedColor(holder.itemView.getContext(), R.b.colorBackgroundPrimary));
        
        AvatarUtils avUtil = new AvatarUtils(friendUser);
        if (friendUser.getAvatar() != null) {
            Utils.threadPool.execute(() -> {
                byte[] decodedString = Base64.decode(avUtil.toBase64(), Base64.DEFAULT);
                Bitmap bitMap = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
                Utils.mainThread.post(() -> {
                    holder.icon.setImageBitmap(AvatarUtils.makeCircle(bitMap));
                });
            });
            holder.iconText.setVisibility(View.GONE);
        } else {
            holder.icon.setVisibility(View.GONE);
        }

        holder.name.setText(friendUser.getUsername());
        if(holder.nick != null) holder.nick.setVisibility(View.GONE);
    }

    public void onClick(Context ctx, int position) {
        var friend = friends.get(position);
        try {
            final FragmentActivity activity = (FragmentActivity) ctx;
            WidgetUserSheet.Companion.show(friend.getUserId(), activity.getSupportFragmentManager());
        } catch (Throwable e) {  logger.error("Error opening user sheet", e);}
    }
}

File: Sessions/src/main/java/xyz/wingio/plugins/sessions/Adapter.kt
package xyz.wingio.plugins.sessions

import android.annotation.SuppressLint
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.view.*
import android.widget.ImageView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.aliucord.PluginManager
import com.aliucord.plugins.R
import com.discord.utilities.color.ColorCompat
import com.discord.utilities.time.ClockFactory
import com.discord.utilities.time.TimeUtils
import xyz.wingio.plugins.utils.Utils

class Adapter(var data: MutableList<Session>) :
    RecyclerView.Adapter<Adapter.ViewHolder?>() {

    var onLogoutClick: (sessionId: String) -> Unit = {}

    class ViewHolder(private val adapter: Adapter, val item: SessionCard) : RecyclerView.ViewHolder(item)

    override fun getItemCount() = data.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(this, SessionCard(parent.context))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val session = data[position]
        val mobileOsList = listOf("Android", "iOS")

        holder.item.logOutBtn.apply {
            visibility = View.VISIBLE
            setOnClickListener {
                onLogoutClick(session.idHash)
            }
        }

        holder.item.title = "${session.clientInfo.os ?: "Unknown"} ${if(session.clientInfo.platform != null) " · ${session.clientInfo.platform}" else ""}"
        holder.item.location = session.clientInfo.location ?: "Unknown Location"

        holder.item.timestamp.visibility = View.VISIBLE
        holder.item.timestampText = Utils.getTimestampString(session.approxLastUsedTime.g())

        holder.item.isMobile = mobileOsList.contains(session.clientInfo.os ?: "Unknown")
    }

    fun updateData(newData: List<Session>?) {
        newData?.let {
            data.clear()
            addData(newData)
        }
    }

    fun addData(data: List<Session>?) {
        this.data.addAll(data!!)
        notifyDataSetChanged()
    }

}